feat: add mobile credential online presentation sample app (#167)
mattrglobal/sample-apps: mobile-credential-online-presentation/README.md

> [!NOTE]
> This is a tech preview. If you are interested in in trialling this capability, please reach out via [https://mattr.global/contact-us](https://mattr.global/contact-us).

# Mobile Credential Online Presentation Sample App

> This sample app demonstrates the usage of the Web Verifier SDK to present mobile credentials online on the same device and across devices.

Reviewer: I think the term they have settled on was Verifier Web SDK

Author: Fixed. Also adjusted some SDK references to use the full name for consistency.

> This sample app demonstrates the usage of the Web Verifier SDK to present mobile credentials online on the same device and across devices.

Reviewer: ```suggestion
This sample app demonstrates using the Web Verifier SDK for online presentation of [Mobile Credentials](https://learn.mattr.global/docs/profiles/mobile) via either a same-device or cross-device workflow.
```

## Prerequisites

### MATTR verifier tenant configuration

> 

Reviewer: We can probably link to the tech-preview API Ref instead of detailing all of this here? 

https://online-presentations-tech-preview.redoc.ly/tag/Mobile-Credentials-Verification#operation/putVerifierConfiguration

Then just say something like

You must create a Verifier configuration on your MATTR VII verifier tenant. Refer to the API reference for a complete description of this configuration and different settings you can control.

Also might be worth calling out that they need to sign up for a trial tenant if they don't have access to one?

Author: Fixed in https://github.com/mattrglobal/sample-apps/pull/167/commits/776c8c6240f228f1fbf937a92be937ecf460594f.

I did not specifically call out signing up for a trial tenant, because the banner on top of the README calls out getting in touch anyways. Trial tenants would not have online presentation enabled by default.

Your MATTR verifier tenant needs to be made aware of the public domain of the app.
You can also provide additional configuration, e.g. restricting the supported modes or customizing the iframe for the cross-device flow.

A sample request to configure your verifier tenant using `curl` looks like this:

```bash
curl -X PUT https://<YOUR_TENANT_DOMAIN>/v2/presentations/configuration \
-H "Authorization: Bearer <YOUR_ACCESS_TOKEN>" \
-H "Content-Type: application/json" \
-d '{
    "supportedMode": "all",
    "domains": ["<PUBLIC_DOMAIN_OF_THE_APP>"],
    "redirectUris": [
        "https://<PUBLIC_DOMAIN_OF_THE_APP>"
    ],
    "display": {
        "bodyText": "Please scan the QR code to the right to provide information required for this interaction.",
        "logoImage": {
            "url": "<YOUR_LOGO_URL>",
            "altText": "<YOUR_LOGO_ALT_TEXT>"
        },
        "headerText": "Verify your age before you continue",
        "primaryColorHex": "#000000"
    },
    "resultAvailableInFrontChannel": true
}'
```

### Making the app publically available

> The app needs to be publically available for the online presentation to work.

Reviewer: ```suggestion
The app needs to be publicly available for online presentation to work. You can either:
- Use a hosting service of your choice.
- Use cloudflare or ngrok tunnels to make your locally running app publicly available.
```

You can either use a hosting service of your choice, or use cloudflare or ngrok tunnels to expose your locally running app publically.

> [!IMPORTANT]
> If you use a tunneling service, make sure to

> > * update the redirect URIs in your verifier tenant configuration (see above), and to

Reviewer: ```suggestion
> * Update the [redirect URIs](https://online-presentations-tech-preview.redoc.ly/tag/Mobile-Credentials-Verification#operation/putVerifierConfiguration!path=redirectUris&t=request) in your verifier tenant configuration (see above).
```

> > * access the app yourself via the public domain.

Reviewer: ```suggestion
> * Access the app via the public domain when you are interacting with it.
```

>

> > The credential request to the verifier tenant validates the origin of the request and the request will fail when it is being made from `localhost`.

Reviewer: ```suggestion
> The origin of the credential request is validated by the verifier tenant and the request will fail when it is being made from `localhost`.
```

### Credential holding in the wallet

The default credential query requests a credential of doctype `org.iso.18013.5.1.mDL`, but you can change the credential query in the sample app directly.

> You need to ensure your wallet holds a credential that can fulfil the credential query you are using.

Reviewer: ```suggestion
Make sure your wallet holds a credential that matches the credential query you are configuring.
```

## SDK Guide

### Automatic mode detection

> If you don't provide a `mode` in the `RequestCredentialsOptions`, the SDK will determine the mode automatically by checking the user agent of the device in use.

Reviewer: ```suggestion
If you don't provide a `mode` in the `RequestCredentialsOptions`, the SDK will determine the mode automatically by checking the user agent of the device in use:
```

> Mobile devices will use the same device flow, all other devices the cross device flow.

Reviewer: ```suggestion
* Mobile devices will use the same-device flow.
* Other devices will use the cross-device flow.
```

> ### Same device presentation

Reviewer: ```suggestion
### Same-device workflows
```

> When request credentials in the same device flow, the user will be redirected to a wallet to fulfil the credential request.

Reviewer: ```suggestion
When credentials are requested in the same-device flow, the user is redirected to a wallet to fulfil the request.
```

> On completion of the request in the wallet, the user is redirected back to the browser to the redirect URI from the credential request.

Reviewer: ```suggestion
On completion of the request in the wallet, the user is redirected back to the browser using the redirect URI configured in the credential request.
```

> The `handleRedirectCallback` function from the SDK will extract the `response_code` from the URL, and retrieve the presentation results.

Reviewer: ```suggestion
The SDK's [`handleRedirectCallback`](https://api-reference-sdk.mattr.global/verifier-sdk-web/preview/functions/handleRedirectCallback.html) function extracts the `response_code` from the URL and retrieve the presentation results.
```

> [!NOTE]

> > For the SDK to be able to retrieve the results on rendering the page the user is redirected to, the SDK needs to have the verifier tenant URL accessible.

Reviewer: ```suggestion
> For the SDK to retrieve the results when rendering the page the user is redirected to, the verifier tenant URL must be accessible to the SDK.
```

> > The simplest way is to provide it as an environment variable when running the app.

Reviewer: ```suggestion
> The recommended way to achieve this is to provide the URL as an environment variable when running the app.
```

> Alternatively, you can persist the verifier tenant URL in `localStorage` when making the credential request, and retrieve it on the redirected tab.

> ### Cross device presentation

Reviewer: ```suggestion
### Cross-device workflows
```

> In the cross device flow, you don't need to invoke the `handleRedirectCallback`, but you can directly pass an `onComplete` and an `onFailure` handler to the `crossDeviceCallback` option in the credential request.

Reviewer: ```suggestion
In the cross-device flow you can pass an `onComplete` and an `onFailure` handler directly to the `crossDeviceCallback` option in the credential request. You do not need to invoke the `handleRedirectCallback` function. 
```

## Running the app

This is a NextJS app, using the App Router.
You can configure your verifier tenant URL by setting the `NEXT_PUBLIC_API_BASE_URL` environment variable, e.g. by adding
```
NEXT_PUBLIC_API_BASE_URL=<YOUR_TENANT_URL>
```

> to a `.env.local` file to the root of the project.

Reviewer: ```suggestion
to a `.env.local` file at the root of the project.
```

> This app only consists of a single page, and the main code on how to use the Web Verifier SDK can be found in `src/app/page.tsx`.

Reviewer: ```suggestion
This is a single page app. The main code using the Web Verifier SDK can be found in `src/app/page.tsx`.
```

Author: "single page app" (i.e. SPA) means something specific in this context. And this app is technically not a SPA. I'll improve the sentence similar to your suggestion, though.

Author: Fixed in https://github.com/mattrglobal/sample-apps/pull/167/commits/776c8c6240f228f1fbf937a92be937ecf460594f.

To run the app, call the following commands in your terminal:

```bash
# Install dependencies
npm install

# Run the app locally
npm run dev
```

> [!NOTE]

> > When presenting a mobile credential with the same device flow, the SDK is performing a side effect after the redirect by loadnig the session results.

Reviewer: ```suggestion
> When presenting a Mobile Credential via a same-device workflow, the SDK is performing a side-effect after the redirect by loading the session results.
```

> > When presenting a mobile credential with the same device flow, the SDK is performing a side effect after the redirect by loadnig the session results.

Reviewer: I think that's also configurable?

Author: You can decide to do it differently, e.g. you could render a button for the user to click to retrieve the results. For the sample app, we opted to doing it automatically without any user interaction, which makes this note necessary.

I'll try to highlight that it is an implementation choice in this sample app.

Author: Fixed in https://github.com/mattrglobal/sample-apps/pull/167/commits/776c8c6240f228f1fbf937a92be937ecf460594f.

> > This conflicts with Reacts "Strict Mode", which is intentionally running every every effect twice.

Reviewer: ```suggestion
> This conflicts with React's _Strict Mode_, which is intentionally running every effect twice.
```

> > If you expierence any issue with disappearing results, check whether "Strict mode" is disabled in your NextJS configuration (`next.config.mjs`).

Reviewer: ```suggestion
> If you experience any issues with disappearing results, check whether _Strict mode_ is disabled in your NextJS configuration (`next.config.mjs`).
```
